vaadin-date-picker-flow/src/main/resources/META-INF/resources/frontend/datepickerConnector.js

```diff
@@ -78,69 +132,176 @@
                 /* create test-string where to extract parsing regex */
-                let testDate = new Date(Date.UTC(datepicker.$connector.yearPart.initial, datepicker.$connector.monthPart.initial - 1, datepicker.$connector.dayPart.initial));
+                let testDate = new Date(Date.UTC(parseInt(yearPart.initial), parseInt(monthPart.initial) - 1, parseInt(dayPart.initial)));
                 let testString = cleanString(testDate.toLocaleDateString(locale, { timeZone: 'UTC' }));
-                datepicker.$connector.parts.forEach(function (part) {
+                parts.forEach(function (part) {
                     part.index = testString.indexOf(part.initial);
                 });
                 /* sort items to match correct places in regex groups */
-                datepicker.$connector.parts.sort(FlowDatePickerPart.compare);
+                parts.sort(FlowDatePickerPart.compare);
                 /* create regex
-                * regex will be the date, so that:
-                * - day-part is '(\d{1,2})' (1 or 2 digits),
-                * - month-part is '(\d{1,2})' (1 or 2 digits),
-                * - year-part is '(\d{4})' (4 digits)
-                *
-                * and everything else is left as is.
-                * For example, us date "10/20/2010" => "(\d{1,2})/(\d{1,2})/(\d{4})".
-                *
-                * The sorting part solves that which part is which (for example,
-                * here the first part is month, second day and third year)
-                *  */
-                datepicker.$connector.regex = testString.replace(/[-[\]{}()*+?.,\\^$|#\s]/g, '\\$&').replace(datepicker.$connector.dayPart.initial, "(\\d{1,2})").replace(datepicker.$connector.monthPart.initial, "(\\d{1,2})").replace(datepicker.$connector.yearPart.initial, "(\\d{4})");
-
-                datepicker.i18n.formatDate = tryCatchWrapper(function (date) {
+                 * regex will be the date, so that:
+                 * - day-part is '(\d{1,2})' (1 or 2 digits),
+                 * - month-part is '(\d{1,2})' (1 or 2 digits),
+                 * - year-part is '(\d{1,4})' (1 to 4 digits)
+                 *
+                 * and everything else is left as is.
+                 * For example, us date "10/20/2010" => "(\d{1,2})/(\d{1,2})/(\d{1,4})".
+                 *
+                 * The sorting part solves that which part is which (for example,
+                 * here the first part is month, second day and third year)
+                 */
+                datepicker.$connector.regex = testString.replace(/[-[\]{}()*+?.,\\^$|#\s]/g, '\\$&')
+                    .replace(dayPart.initial, "(\\d{1,2})")
+                    .replace(monthPart.initial, "(\\d{1,2})")
+                    .replace(yearPart.initial, "(\\d{1,4})");
+
+                /**
+                 * @param {DateHash} date
+                 * @returns {number}
+                 */
+                let indexOfYearForFormattedDate = function (date) {
```

Comment:
A new helper method to check at which position in a formatted date string does the year begin (this location depends on used locale). It's needed for being able to prefix the year in a formatted string with zeros to make it 4 digits (in case the year is less than 1000) since there is no other way to enforce a year to be formatted with 4 digits when formatting with `Date.toLocaleDateString()` (in all locales).